controllers/internal/utils/dns.go

```diff
@@ -20,30 +20,99 @@ package utils
 	"fmt"
 	"sort"
+	"strings"
 
 	"github.com/lixiangzhong/dnsutil"
+	"github.com/miekg/dns"
 )
 
+type DNSServer struct {
+	Host string
+	Port int
+}
+
 // Dig retrieves list of tuple <IP address, A record > from edge DNS server for specific FQDN
 func Dig(edgeDNSServer, fqdn string) ([]string, error) {
-	var dig dnsutil.Dig
 	if edgeDNSServer == "" {
 		return nil, fmt.Errorf("empty edgeDNSServer")
 	}
-	err := dig.SetDNS(edgeDNSServer)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)", edgeDNSServer, err)
-		return nil, err
+	dnsServer := DNSServer{
+		Host: edgeDNSServer,
+		Port: 53,
+	}
+	return DigWithFallback([]DNSServer{dnsServer}, fqdn)
+}
+
+func DigWithFallback(edgeDNSServers []DNSServer, fqdn string) ([]string, error) {
+	var dig dnsutil.Dig
+	if len(edgeDNSServers) == 0 {
+		return nil, fmt.Errorf("empty edgeDNSServers, provide at least one")
 	}
-	a, err := dig.A(fqdn)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
-		return nil, err
+	for i := 0; i < len(edgeDNSServers); i += 2 {
+		if edgeDNSServers[i].Host == "" {
+			continue
+		}
+		err := dig.SetDNS(OneToString(edgeDNSServers[i]))
+		if err != nil {
+			err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)",
+				OneToString(edgeDNSServers[i]), err)
+			return nil, err
+		}
+		if len(edgeDNSServers) > i+1 {
+			backupDNSServer := edgeDNSServers[i+1]
+			if backupDNSServer.Host == "" {
+				return nil, fmt.Errorf("empty edgeDNSServer")
+			}
+			err := dig.SetBackupDNS(OneToString(backupDNSServer))
+			if err != nil {
+				err = fmt.Errorf("dig error: can't set dns backup server (%s) with error(%s)",
+					OneToString(backupDNSServer), err)
+				return nil, err
+			}
+		}
+		a, err := dig.A(fqdn)
+		if err != nil {
+			if len(edgeDNSServers) > i+2 { // there are still fallback dns servers listed
+				continue
+			} else {
+				err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
+				return nil, err
+			}
+		}
+		var IPs []string
+		for _, ip := range a {
+			IPs = append(IPs, fmt.Sprint(ip.A))
+		}
+		sort.Strings(IPs)
+		return IPs, nil
 	}
-	var IPs []string
-	for _, ip := range a {
-		IPs = append(IPs, fmt.Sprint(ip.A))
+	return nil, fmt.Errorf("dig error: all dns servers were tried and none of them were able to resolve")
+}
+
+func Exchange(m *dns.Msg, edgeDNSServers []DNSServer) (*dns.Msg, error) {
+	if len(edgeDNSServers) == 0 {
+		return nil, fmt.Errorf("empty edgeDNSServers, provide at least one")
+	}
+	var msg *dns.Msg
+	var err error
+	for _, ns := range edgeDNSServers {
+		msg, err = dns.Exchange(m, OneToString(ns))
+		if err != nil {
+			continue
+		}
+		return msg, err
+	}
+	return nil, fmt.Errorf("exchange error: all dns servers were tried and none of them were able to resolve, err: %s", err)
+}
+
+func OneToString(s DNSServer) string {
+	return fmt.Sprintf("%s:%v", s.Host, s.Port)
+}
+
+func AllToString(s []DNSServer) string {
```

Comment:
Is used by depresolver only, could be private within depresolver package. 

not necesary, but consider:
```go
type DNSList []DNSServer

func (l DNSList) StringSlice() (s []string) {
	for _, x := range l {
		s = append(s, x.String())
	}
	return
}
// strings.Join(DNSList(xx).StringSlice(),",")

or

func (l DNSList) String() (s string) {
	var a []string
	for _, x := range l {
		a = append(a, x.String())
	}
	return strings.Join(a,",")
}

// fmt.Println(DNSList(xx))
```